From f03f0796d114f0952ca5011888c9dd992a70afd9 Mon Sep 17 00:00:00 2001
From: yangbingqian <bqyang720@gmail.com>
Date: Wed, 22 Mar 2017 18:41:15 +0800
Subject: [PATCH 020/296] FATAL EXCEPTION IN SYSTEM PROCESS: android.ui

when 3rd party app put an extra with a bad serial object
to start a service or sendbroadcast to system_server.
It will cause system_server crash because of
classNotFoundException.

A test demo apk will cause system_server crash.
It really affact the system stability

Intent intent = new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS);
intent.putExtra(Intent.EXTRA_REFERRER, new serial()
/*a Serializable class object*/);
sendBroadcast(intent);

Test: use the test apk attached in the issue below, system_server will
crash every time. And it'll be OK with this CL.
---
 core/java/android/os/BaseBundle.java | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/core/java/android/os/BaseBundle.java b/core/java/android/os/BaseBundle.java
index 5312dcaeb50..aa32217bf61 100644
--- a/core/java/android/os/BaseBundle.java
+++ b/core/java/android/os/BaseBundle.java
@@ -278,6 +278,13 @@ public class BaseBundle {
             } else {
                 throw e;
             }
+        } catch (RuntimeException e) {
+            if (sShouldDefuse && (e.getCause() instanceof ClassNotFoundException)) {
+                Log.w(TAG, "Failed to parse Bundle, but defusing quietly", e);
+                map.erase();
+            } else {
+                throw e;
+            }
         } finally {
             mMap = map;
             if (recycleParcel) {
-- 
2.11.0

